libcontainer/seccomp/enosys_hotfix_linux.go

```diff
@@ -0,0 +1,547 @@
+				continue
+			}
+			if sysno > largestSyscall {
+				largestSyscall = sysno
+			}
+		}
+		lastSyscalls[nativeArch][arch] = largestSyscall
+	}
+
+	// A jump-table for each nativeArch used to generate the initial
+	// conditional jumps -- measured from the *END* of the program so they
+	// remain valid after prepending to the tail.
+	archJumpTable := map[nativeArch]uint32{}
+
+	// Generate our own -ENOSYS rules for each architecture. They have to be
+	// generated in reverse (prepended to the tail of the program) because the
+	// JumpIf jumps need to be computed from the end of the program.
+	programTail := []bpf.Instruction{
+		// Fall-through rules jump into the filter.
+		bpf.Jump{Skip: 1},
+		// Rules which jump to here get -ENOSYS.
+		bpf.RetConstant{Val: uint32(C.C_ACT_ERRNO_ENOSYS)},
+	}
+
+	// Generate the syscall -ENOSYS rules.
+	for nativeArch, maxSyscalls := range lastSyscalls {
+		// The number of instructions from the tail of this section which need
+		// to be jumped in order to reach the -ENOSYS return. If the section
+		// does not jump, it will fall through to the actual filter.
+		baseJumpEnosys := uint32(len(programTail) - 1)
+
+		// Add the load instruction for the syscall number -- we jump here
+		// directly from the arch code so we need to do it here. Sadly we can't
+		// share this code between architecture branches.
+		section := []bpf.Instruction{
+			// load [0]
+			bpf.LoadAbsolute{Off: 0, Size: 4},
+		}
+
+		switch len(maxSyscalls) {
+		case 1:
+			// Get the only syscall in the map.
+			var sysno libseccomp.ScmpSyscall
+			for _, no := range maxSyscalls {
+				sysno = no
+			}
+
+			// The simplest case just boils down to a single jgt instruction,
+			// with special handling if baseJumpEnosys is larger than 255 (and
+			// thus a long jump is required).
+			if baseJumpEnosys <= 255 {
+				section = append(section, []bpf.Instruction{
+					// jgt [syscall],[baseJumpEnosys]
+					bpf.JumpIf{Cond: bpf.JumpGreaterThan, Val: uint32(sysno), SkipTrue: uint8(baseJumpEnosys)},
+				}...)
+			} else {
+				section = append(section, []bpf.Instruction{
+					// jle [syscall],1
+					bpf.JumpIf{Cond: bpf.JumpLessOrEqual, Val: uint32(sysno), SkipTrue: 1},
+					// ja [baseJumpEnosys]
+					bpf.Jump{Skip: baseJumpEnosys},
+				}...)
+			}
+		case 2:
+			// x32 and x86_64 are a unique case, we can't handle any others.
+			if uint32(nativeArch) != uint32(C.C_AUDIT_ARCH_X86_64) {
+				return nil, errors.Errorf("unknown architecture overlap on native arch %#x", nativeArch)
+			}
+
+			x32sysno, ok := maxSyscalls[libseccomp.ArchX32]
+			if !ok {
+				return nil, errors.Errorf("missing %v in overlapping x86_64 arch: %v", libseccomp.ArchX32, maxSyscalls)
+			}
+			x86sysno, ok := maxSyscalls[libseccomp.ArchAMD64]
+			if !ok {
+				return nil, errors.Errorf("missing %v in overlapping x86_64 arch: %v", libseccomp.ArchAMD64, maxSyscalls)
+			}
+
+			// The x32 ABI indicates that a syscall is being made by an x32
```

Comment:
Who actually uses x32?

Reply:
No idea, but the behaviour would be incorrect if we didn't implement it (all syscalls for x32 binaries would give `-ENOSYS`) -- Docker (and everyone else's) libseccomp configuration is such that x32 syscall rules are always included. If we want to drop x32 support we should drop it from the profiles.

Reply:
I think this is resolved now -- we need to keep x32 support for as long as we include it in our profiles by default.